DEVELOPMENT-GUIDE.md

```diff
@@ -0,0 +1,86 @@
+        When Known Answer Tests from standards are available, we must use them. (Some older tests do not yet meet this standard but compensate with compatibility tests.)
+    2. Cryptographic error cases must be checked.
+        (While useful, API error cases should be checked but for simple type, null, and related errors, it is acceptable to miss some.)
+    3. Different call patterns must be checked
+        Many cryptographic APIs can be called in different ways. These must all be checked.
+	  ([`MessageDigest` example](https://github.com/corretto/amazon-corretto-crypto-provider/blob/develop/src/com/amazon/corretto/crypto/provider/Utils.java#L297))
+5. ACCP must be fast.
+    This is one of the primary purposes of ACCP.
+    Benchmark your code, find bottlenecks, fix them.
+6. Code should be *obviously* correct.
+    A developer should be able to look at an implementation and *know* it is correct with minimal reasoning or justification.
+    By implication, this means that you shouldn't be clever.
+    This principle sometimes needs to be sacrificed to support higher priority tenets. When this happens, we must do the following:
+    1. Isolate complexity. (So that only a few methods or a single file is hard to understand.)
+    2. Test to prove correctness. (While we must always do this, it is even more critical here.)
+    3. Comment to explain exactly what is intended and, when appropriate, why a particular technique was choseen.
+    Examples: ([ConstantTime](https://github.com/corretto/amazon-corretto-crypto-provider/blob/develop/src/com/amazon/corretto/crypto/provider/ConstantTime.java) and its [tests](https://github.com/corretto/amazon-corretto-crypto-provider/blob/develop/tst/com/amazon/corretto/crypto/provider/test/ConstantTimeTests.java), [Janitor](https://github.com/corretto/amazon-corretto-crypto-provider/blob/develop/src/com/amazon/corretto/crypto/provider/Janitor.java) and its [tests](https://github.com/corretto/amazon-corretto-crypto-provider/blob/develop/tst/com/amazon/corretto/crypto/provider/test/JanitorTest.java))
+
+# Important and Unique Components
+## Java
+### Janitor
+ACCP never uses [finalizers](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Object.html#finalize()) due to significant performance problems. Since we still need to support Java8 for the foreseable future, we have implemented [Janitor](https://github.com/corretto/amazon-corretto-crypto-provider/blob/develop/src/com/amazon/corretto/crypto/provider/Janitor.java) as a JDK8+ replacement for the newer (since JDK9) [Cleaner](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/ref/Cleaner.html). To avoid circular dependency issues, `Janitor` *MUST NOT* depend on any other ACCP resources (directly or indirectly). It must remain entirely self contained.
```

Comment:
IIRC when I tested Cleaner it had some contention issues that would have made it unusable. It's been a few years though.

Reply:
Note added that we will need to re-evaluate it when the time comes.